gh-89520: IDLE - Make extentions use user's keys, not all defaults

CoolCat467 opened the pull request:
I was trying to write an extension for Idle, and was noticing extension keys defined in ~/.idlerc weren't being used, so I looked into it and found that in idlelib.config.idleConf, `GetExtensionKeys`, `__GetRawExtensionKeys`, and `GetKeyBinding` would never check for user keys, always assuming default keys have everything. This is not always true. In a system with multiple users, you would not want to modify the default config file for a custom extension, as that is used for everyone. This pull request is changing aforementioned functions to also check user config files, and overwrite values acquired from the default config file if the user config redefined one.

Edit:
The especially interesting part is that extensions are loaded anyways if they have an entry in the user's IDLE configuration file, but their keybinds are only loaded properly if they are defined in the root configuration file at the moment. This pull request is fixing this difference and allows extension keybinds to be loaded from the user's IDLE configuration file, making user extensions work just as well as root extensions.

<!-- gh-issue-number: gh-89520 -->
* Issue: gh-89520
<!-- /gh-issue-number -->

CoolCat467 committed:
Make extentions use user's keys, not all defaults

I was trying to write an extension for Idle, and was noticing extension keys defined in ~/.idlerc weren't being used, so I looked into it and found `GetExtensionKeys`, `__GetRawExtensionKeys`, and `GetKeyBinding` would never check for user keys, always assuming default keys have everything. This is not always true. In a system with multiple users, you would not want to modify the default config file for a custom extension, as that is used for everyone. This pull request is changing aforementioned functions to also check user config files, and overwrite values acquired from the default config file if the user config redefined one.

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).



## CLA Missing

Our records indicate the following people have not signed the CLA:

@CoolCat467

For legal reasons we need all the people listed to sign the CLA before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.

If you have recently signed the CLA, please wait at least one business day
before our records are updated.


You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

CoolCat467 committed:
Fix config and dialog to use user config extentions

Config saving and loading also weren't working quite right for extensions.

CoolCat467 committed:
Merge branch 'python:main' into patch-1

CoolCat467 committed:
Remove debug print I forgot

CoolCat467 committed:
Remove debug comment out lines

CoolCat467 committed:
Merge branch 'python:main' into patch-1

CoolCat467 committed:
Merge branch 'python:main' into patch-1

CoolCat467 committed:
Merge branch 'python:main' into patch-1

JelleZijlstra commented on Lib/idlelib/config.py:
```diff
@@ -192,7 +193,7 @@ def GetUserCfgDir(self):
                         pass
                 userDir = '~'
-        if userDir == "~": # still no path to home!
+        else: # still no path to home!
```
This changes behavior because line 194 may change `userDir`.

CoolCat467 replied:
Reverted change

JelleZijlstra commented on Lib/idlelib/config.py:
```diff
@@ -755,7 +777,8 @@ def LoadCfgFiles(self):
             self.defaultCfg[key].Load()
-            self.userCfg[key].Load() #same keys
+        for key in self.userCfg:
+            self.userCfg[key].Load() #not nessicerally same keys, could be more
```
```suggestion
            self.userCfg[key].Load() #not necessarily same keys, could be more
```

CoolCat467 committed:
Fix spelling

Co-authored-by: Jelle Zijlstra <jelle.zijlstra@gmail.com>

CoolCat467 committed:
Revert previous change

CoolCat467 committed:
Merge branch 'python:main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Add trailing newline

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'python:main' into patch-1

CoolCat467 committed:
Simplify and update from `idleuserextend`'s testing

CoolCat467 committed:
Revert always save changes

CoolCat467 committed:
Merge branch 'python:main' into patch-1

JelleZijlstra commented on Lib/idlelib/config.py:
```diff
@@ -192,7 +193,7 @@ def GetUserCfgDir(self):
                         pass
                 userDir = '~'
-        if userDir == "~": # still no path to home!
+        if userDir == '~': # still no path to home!
```
Avoid making unnecessary changes that are unrelated to your functional change

CoolCat467 replied:
Reverted unnecessary formatting changes in dc46fb7

JelleZijlstra reviewed (commented):
Some tests are failing; please fix those.

I don't know enough about IDLE to review this diff's substantive change, so I'll leave the PR to Terry.

CoolCat467 committed:
Revert unnecessary formatting changes

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
`user_list` -> `user`, as is no longer a list but a set.

CoolCat467 commented:
> Some tests are failing; please fix those.

Fixed failing tests as of 016ce89

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

JelleZijlstra commented:
As mentioned above, I'm not able to provide a full review on this PR.

CoolCat467 commented:
@terryjreedy Ping, it's been 3 years and I think my newest approach will work well. I've been testing a version of what this PR does [as it's own module](https://github.com/CoolCat467/idleuserextend) that has to do some nasty monkeypatching stuff, but it allows IDLE to successfully load extensions from user configuration files, and I would like to see this become something standard.

terryjreedy reviewed (commented):
These comments are very preliminary.  And when I wrote the first 2, I had forgotten that this PR intentionally changes a config rule.  But its late, so I am submitting them as they are to give you some feedback.

I am a bit more favorable toward this than I was on the issue, but no promises yet.

If I were to approve, new tests would be needed, modeled on existing tests that avoid reading/writing the user files on disk by loading the in-memory config dicts.  zzdummy.py, or a copy in idle_test, might be usable.

CoolCat467 commented on Lib/idlelib/editor.py:
```diff
@@ -904,11 +905,12 @@ def RemoveKeybindings(self):
         self.mainmenu.default_keydefs = keydefs = idleConf.GetCurrentKeySet()
         for event, keylist in keydefs.items():
-            self.text.event_delete(event, *keylist)
+            with contextlib.suppress(ValueError):
+                self.text.event_delete(event, *keylist)
```
I do not recall immediately what the issue was, I can look into that.

CoolCat467 replied:
Looking into it more, this particular suppress is useless, there is no errors. The following one does have errors though, erroring whenever it attempts to remove an event defined from an extension that was loaded from user configuration files.
```console
Partial Traceback:
    editwin.text.event_delete(event, *keylist)
  File "/usr/lib/python3.12/idlelib/multicall.py", line 391, in event_delete
    self.__binders[triplet[1]].unbind(triplet, func)
  File "/usr/lib/python3.12/idlelib/multicall.py", line 234, in unbind
    doit()
  File "/usr/lib/python3.12/idlelib/multicall.py", line 232, in <lambda>
    doit = lambda: self.bindedfuncs[triplet[2]][triplet[0]].remove(func)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: list.remove(x): x not in list
```
I will look deeper into this.

CoolCat467 replied:
Finished, turns out this error will never happen either, I am only having this error because my extension that monkeypatches IDLE is loaded after the builtin extensions have loaded, which leads to some issues later. This issue should never happen in a copy of IDLE that loads user extensions from the get go.

Removed suppression in 024095903e

CoolCat467 committed:
Revert noted strictly format change

CoolCat467 commented on Lib/idlelib/configdialog.py:
```diff
@@ -1960,23 +1960,33 @@ def create_page_extensions(self):
+                opt_name
+                for opt_name in opt_list
+                if opt_name.startswith('enable')
+            ]
```
Reverted as of 568cd80d14

CoolCat467 commented on Lib/idlelib/config.py:
```diff
@@ -757,7 +792,8 @@ def LoadCfgFiles(self):
         for key in self.defaultCfg:
             self.defaultCfg[key].Load()
-            self.userCfg[key].Load() #same keys
+        for key in self.userCfg:
+            self.userCfg[key].Load()
```
This change allows keys that are not in the default configuration to be loaded from user configuration. This means sections like extensions that might not be in the default config can be loaded.

CoolCat467 committed:
Remove suppression for errors that will never happen

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Add test for ZzDummy being loaded in extensions list

CoolCat467 committed:
Don't require enabled

CoolCat467 committed:
Add tests for reading extension keys

CoolCat467 commented:
> These comments are very preliminary. And when I wrote the first 2, I had forgotten that this PR intentionally changes a config rule. But its late, so I am submitting them as they are to give you some feedback.
> 
> If I were to approve, new tests would be needed, modeled on existing tests that avoid reading/writing the user files on disk by loading the in-memory config dicts. zzdummy.py, or a copy in idle_test, might be usable.

@terryjreedy Ping, I have reviewed your feedback and have made noted changes, and I've added a few tests that try to capture the fact that and extension that is not enabled will not have it's keybinds bound but that one that is enabled will, as well as making sure extensions that are enabled entirely from user configuration files will have their keybinds found successfully. To do this I made a copy of `test_zzdummy` (`test_zzdummy_user`) and changed both of them to test aforementioned behavior.

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 committed:
Merge branch 'main' into patch-1

CoolCat467 commented:
@gpshead Would you mind looking at this one as well? It's been nearly two years since I've last heard from terry about this.

gpshead committed:
Merge branch 'main' into patch-1

gpshead committed:
gh-89520: IDLE - Replace ChainMap misuse with set union in config.py

ChainMap is designed for Mapping types, not lists. GetOptionList()
returns a list of strings, so passing these to ChainMap only worked
by accident via CPython's dict.fromkeys() in ChainMap.__iter__.

Replace with set() union, matching the pattern already used in
GetExtensionKeys. The actual user-over-default prioritization is
handled by GetOption(), not by the iteration order.

gpshead committed:
gh-89520: IDLE - Use self instead of module-level idleConf in GetExtensionKeys

GetExtensionKeys is a method on IdleConf. Using the module-level
singleton idleConf instead of self is inconsistent with every other
method in the class.

gpshead committed:
gh-89520: IDLE - Fix default value logic in configdialog load_extensions

Fix two issues in load_extensions:

1. def_str (the "default" value) was sourced from user config when the
   option existed there. This meant opening the extensions dialog could
   silently revert user customizations, since the user's value would be
   treated as the default and then removed. Now prefers default config,
   falling back to user config only for user-only extensions.

2. The value lookup was skipping user config for default-only options.
   Restore the original behavior of always trying user config first
   (which returns def_obj as fallback when the option is absent).

gpshead committed:
gh-89520: IDLE - Update NEWS entry to gh-issue format and fix content

Rename from bpo-45357 to gh-issue-89520 format. Fix content to remove
incorrect mention of GetKeyBinding (not changed in this PR).

gpshead committed:
gh-89520: IDLE - Fix test_exists_user cleanup to properly remove section

read_dict is additive and doesn't remove sections. Setting enable to
False still leaves the ZzDummy section in user config. Use
remove_section to fully restore the original empty state.

gpshead committed:
gh-89520: IDLE - Factor out shared test code from zzdummy test files

Create ZZDummyMixin in test_zzdummy.py containing shared test methods
(checklines, test_init, test_reload, test_z_in_event, test_z_out_event,
test_roundtrip). Both ZZDummyTest classes now inherit from the mixin,
eliminating ~90 lines of duplication from test_zzdummy_user.py.

gpshead committed:
gh-89520: IDLE - Wrap long lines in zzdummy test files

gpshead committed:
address terryjreedy's comment, no need for the split loop

gpshead committed:
gh-89520: IDLE - Preserve public API parameter names in config.py

Restore the original extensionName parameter name on GetExtensionKeys
and GetExtensionBindings to avoid breaking callers using keyword
arguments.

CoolCat467 commented:
Thank you so much! I've been waiting to get this merged for five years now, glad this is finally done.
